drivers/syslog/syslog_filechannel.c

```diff
@@ -85,28 +86,44 @@ static void log_rotate(FAR const char *log_file)
       return;
     }
 
-  /* Construct the backup file name. */
+  /* Rotated file names. */
 
-  backup_file = malloc(strlen(log_file) + 3);
-  if (backup_file == NULL)
+  name_size = strlen(log_file) + 8;
+  rotate_to = kmm_malloc(name_size);
+  rotate_from = kmm_malloc(name_size);
+  if ((rotate_to == NULL) || (rotate_from == NULL))
     {
-      return;
+      goto end;
     }
 
-  sprintf(backup_file, "%s.0", log_file);
+  /* Rotate the logs. */
+
+  for (i = (CONFIG_SYSLOG_FILE_ROTATIONS - 1); i > 0; i--)
+    {
+      snprintf(rotate_to, name_size, "%s.%d", log_file, i);
+      snprintf(rotate_from, name_size, "%s.%d", log_file, i - 1);
+
+      if (access(rotate_to, F_OK) == 0)
+        {
+          remove(rotate_to);
+        }
+
+      rename(rotate_from, rotate_to);
+    }
 
-  /* Delete any old backup files. */
+  snprintf(rotate_to, name_size, "%s.0", log_file);
 
-  if (access(backup_file, F_OK) == 0)
+  if (access(rotate_to, F_OK) == 0)
```

Comment:
let's always remove rotate_to

Reply:
Thinking of it, do we really need the `remove()`?

If `rotate_to` exists, `rename` will replace it. Correct?

Reply:
I think the rename will report error, but you can try.

Reply:
I just tried it, and it works correctly.  
I also followed the code, and indeed it checks explicitly whether the "new" file exists, and if so it deletes it.

Thus the `remove()` here is totally redundant.

Tomorrow morning I will provide an updated version.